File: GC-content.py
# Use re to find DNA sequences in file
import re

# Sequence pattern
seqp = re.compile(r'[ATGC]+')

# Read info pattern
labp = re.compile(r'>(\w+_\d+)')

with open("data/GC-content.fasta") as f:
    rawlst = f.read().split('\n')

# Key : read info line
# Value : Sequence
lst = {}

# Build dict from single line read title
# and multiline read sequence
for i in rawlst:
    if labp.match(i) :
        curr = i.strip('>')
        lst[curr] = ""
    if seqp.match(i):
        lst[curr] += i
        
# Dict with GC-content
# Key = label
# Value = GC-Content
perc = {}

def GCperc(seq):
    """ Calculate GC-Content for a given seq.
    Takes a sequence as string and return ratio of G + C
    """
    n = sum([1 for n in seq if n == ('G') or n == ('C')])
    return n/len(seq) 

# Could have used a map function here...
for label, seq in lst.items():
    perc[label] = GCperc(seq)

# Retrieve label where value is max
m = max(perc, key=perc.get)

# Print formatted as required by challenge
print("{}\n{}".format(m, perc[m]*100))

File: src/main/scala/GCContent.scala
import helpers.{Challenge, FastaParser}

object GCContent extends App {

  Challenge("GCcontent.fasta") provides { data =>

    val DNA = FastaParser.parseDNA(data)

    // Find GC-Content for one string
    def GCContent(s: String): Float = {
      val G = s.count(_ == 'G')
      val C = s.count(_ == 'C')
      (G + C) / s.length.toFloat * 100
    }

    // Build map of (Read ID -> GC-Content of read)
    val contents = DNA.map(x => (x._1.tail, GCContent(x._2)))

    // Find max by value
    val biggest = contents.maxBy(_._2)

    println(biggest._1 + "\n" + biggest._2)

  }

}